build_tools/android/common/benchmark_description.py

```diff
@@ -0,0 +1,280 @@
+#!/usr/bin/env python3
+
+# Copyright 2021 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Utilities for describing Android benchmarks.
+
+This file provides common and structured representation of Android devices,
+benchmark definitions, and benchmark result collections, so that they can be
+shared between different stages of the same benchmark pipeline.
+"""
+
+import json
+import re
+import subprocess
+
+from dataclasses import dataclass
+from typing import Any, Dict, Sequence
+
+__all__ = ["AndroidDeviceInfo", "BenchmarkInfo", "BenchmarkResults"]
+
+# A map for IREE driver names. This allows us to normalize driver names like
+# mapping to more friendly ones and detach to keep driver names used in
+# benchmark presentation stable.
+IREE_DRIVER_NAME_MAP = {
+    "iree_llvmaot": "IREE-Dylib",
+    "iree_vmvx": "IREE-VMVX",
+    "iree_vulkan": "IREE-Vulkan",
+}
+
+
+def execute(args: Sequence[str],
+            capture_output: bool = False,
+            treat_io_as_text: bool = True,
+            verbose: bool = False,
+            **kwargs) -> subprocess.CompletedProcess:
+  """Executes a command."""
+  if verbose:
+    cmd = " ".join(args)
+    print(f"cmd: {cmd}")
+  return subprocess.run(args,
+                        check=True,
+                        capture_output=capture_output,
```

Comment:
`capture_output` was not added until Python 3.7. For backward compatibility, you need to use `stdin=subprocess.PIPE, stdout=subprocess.PIPE`

See a similar helper in https://github.com/google/iree/blob/main/build_tools/docker/utils.py

I think there may also be some complication here because I think `capture_output` with check also blackholes the stderr of the process. Playing around in the python REPL:

```
Python 3.9.2 (default, Feb 28 2021, 17:03:44) 
[GCC 10.2.1 20210110] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import subprocess

>>> subprocess.run(['write_to_stderr_and_fail.sh'])
Does stderr get lost?
CompletedProcess(args=['write_to_stderr_and_fail.sh'], returncode=1)

>>> subprocess.run(['write_to_stderr_and_fail.sh'], check=True)
Does stderr get lost?
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.9/subprocess.py", line 528, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['write_to_stderr_and_fail.sh']' returned non-zero exit status 1.

>>> subprocess.run(['write_to_stderr_and_fail.sh'], capture_output=True, check=True)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.9/subprocess.py", line 528, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['write_to_stderr_and_fail.sh']' returned non-zero exit status 1.

>>> subprocess.run(['write_to_stderr_and_fail.sh'], stderr=subprocess.PIPE, stdout=subprocess.PIPE, check=True)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.9/subprocess.py", line 528, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['write_to_stderr_and_fail.sh']' returned non-zero exit status 1.

>>> subprocess.run(['write_to_stderr_and_fail.sh'], stdout=subprocess.PIPE, check=True)
Does stderr get lost?
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.9/subprocess.py", line 528, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['write_to_stderr_and_fail.sh']' returned non-zero exit status 1.
```

So likely what we actually want is just to set `stdout=subprocess.PIPE`. Also every usage in this file captures the output and immediately get stdout on it, so maybe the helper that makes sense should just return that. Left a separate suggested edit

Reply:
Ah, backward compatibility..